### PR TITLE
Serve user uploads with new UserMediaView

### PR DESCRIPTION
**Pull request by miggol**
This prohibits non-logged-in users from accessing them

### CODE REVIEW

File: main/views.py [hunk 1]
```diff
@@ -449,6 +452,27 @@ def delete(self, request, *args, **kwargs):
         return super(DeleteView, self).delete(request, *args, **kwargs)
 
 
+class UserMediaView(LoginRequiredMixin, generic.View):
+    """Respond with a given user uploaded file if the request
+    is logged in. Makes it so that user uploads are no longer
+    public."""
+
+    def get_response_file(self, filename):
+        filepath = os.path.join(settings.MEDIA_ROOT, filename)
```
Comment (tymees) [inline main/views.py:461]: As noted in the chat, this is dangerous. An attacker can craft a filename that will result in arbitrary file access on the server. (outside MEDIA_ROOT)

It's not a OMG WE ARE DOOMED scenario, we're still behind a LoginRequiredMixin and Apache seems to be filtering out the more simple malicious filenames, but I still really don't want to rely on those protections alone. Especially as the fix is one if statement ;)

Thus, there should be a check for making sure that the resulting filepath is still inside MEDIA_ROOT.

File: fetc/urls.py [hunk 2]
```diff
@@ -20,6 +22,16 @@
...
 
+    # Access user uploads
+    path('media/<str:filename>',
+         UserMediaView.as_view(),
+         name='user_upload'),
+    # And a duplicate path for user media
+    # Use this for testing on servers that rewrite /media/
+    path('view_media/<str:filename>',
+         UserMediaView.as_view(),
+         name='user_upload_2'),
```
Comment (tymees) [inline fetc/urls.py:33]: Is this still necessary?

It feels like a fix for uWSGI's static-map, but we are removing that anyways. In addition, we currently don't really control the urls presented to the user, sooooo...

**Review by tymees — CHANGES_REQUESTED** [review]

File: main/views.py (hunk 1 above)
Comment (tymees) [reply main/views.py:461]: For those interested, it's called a path traversal attack:
https://owasp.org/www-community/attacks/Path_Traversal

File: main/views.py [hunk 3]
```diff
@@ -449,6 +452,27 @@ def delete(self, request, *args, **kwargs):
...
+        try:
```
Comment (Meesch) [inline main/views.py:462]:
```suggestion
if filename in os.listdir(settings.MEDIA_ROOT):  # check if file exists in media root
```

Would this be sufficient sanitization?

**Review by Meesch — COMMENTED** [review]
Other than Ty's comment I have tested it locally and everything works as expected!

**Review by tymees — APPROVED** [review]

**Pull request merged**